bpo-40334: Produce better error messages on invalid targets

lysnikolaou opened the pull request:
The following error messages get produced:
- `cannot delete ...` for invalid `del` targets
- `cannot assign to ...` for invalid targets in for and with
  statements

Additionally a few `cut`s were added in various places before the
invocation of the `invalid_*` rule, in order to speed things
up.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-40334](https://bugs.python.org/issue40334) -->
https://bugs.python.org/issue40334
<!-- /issue-number -->

lysnikolaou committed:
bpo-40334: Produce better error messages on invalid targets

The following error messages get produced:
- `cannot delete ...` for invalid `del` targets
- `... is an illegal 'for' target` for invalid targets in for
  statements
- `... is an illegal 'with' target` for invalid targets in
  with statements

Additionally a few `cut`s were added in various places before the
invocation of the `invalid_*` rule, in order to speed things
up.

lysnikolaou committed:
Fix test_exceptions

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

pablogsal commented on Lib/test/test_syntax.py:
```diff
@@ -148,6 +148,49 @@
+>>> with a as (x, *(y, z.d())): pass
+Traceback (most recent call last):
+SyntaxError: 'function call' is an illegal 'with' target
+
```
These new tests will fail with the oldparser:

```
./python -X oldparser -m test test_syntax
...
test test_syntax failed -- 10 of 108 doctests failed
test_syntax failed

== Tests result: FAILURE ==

1 test failed:
    test_syntax

Total duration: 187 ms
Tests result: FAILURE

```

pablogsal replied:
You need to update the old parser to match these errors as well as it was done for the assignment case

pablogsal replied:
Another option is triying to produce the same errors as the old parser did, i.e. for this code:

```
>>> for [a, b()] in b: pass
```

produce

```
SyntaxError: cannot assign to function call
```

instead of

```
SyntaxError: 'function call' is an illegal 'for' target
```

if possible I would prefer to go with this if you agree :)

lysnikolaou replied:
Yup, I agree it's better that way. Done in https://github.com/python/cpython/pull/20106/commits/1b6cf0b327f024dc53ab548bbfb2bca55064c31c.

lysnikolaou committed:
Produce the same error messages with the old parser for with and for targets

lysnikolaou committed:
Add invalid_group rule

pablogsal commented on Grammar/python.gram:
```diff
@@ -582,19 +589,15 @@ single_subscript_attribute_target[expr_ty]:
-    | a=t_primary '.' b=NAME &del_target_end { _Py_Attribute(a, b->v.Name.id, Del, EXTRA) }
-    | a=t_primary '[' b=slices ']' &del_target_end { _Py_Subscript(a, b, Del, EXTRA) }
+    | a=t_primary '.' b=NAME !t_lookahead { _Py_Attribute(a, b->v.Name.id, Del, EXTRA) }
+    | a=t_primary '[' b=slices ']' !t_lookahead { _Py_Subscript(a, b, Del, EXTRA) }
     | del_t_atom
 del_t_atom[expr_ty]:
-    | a=NAME &del_target_end { _PyPegen_set_expr_context(p, a, Del) }
+    | a=NAME { _PyPegen_set_expr_context(p, a, Del) }
     | '(' a=del_target ')' { _PyPegen_set_expr_context(p, a, Del) }
     | '(' a=[del_targets] ')' { _Py_Tuple(a, Del, EXTRA) }
     | '[' a=[del_targets] ']' { _Py_List(a, Del, EXTRA) }
```
Much better :clap:

pablogsal commented on Grammar/python.gram:
```diff
@@ -666,6 +676,25 @@ invalid_lambda_star_etc:
+    | '(' a=starred_expression ')' {
+        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(
+            a,
+            "can't use starred expression here") }
```
Stylistic nit (feel free to ignore):

```suggestion
    | '(' a=starred_expression ')' {
        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "can't use starred expression here") }
```

pablogsal commented on Parser/pegen/pegen.c:
```diff
@@ -2091,12 +2091,17 @@ _PyPegen_get_invalid_target(expr_ty e)
+            }
+            else {
+                return _PyPegen_get_invalid_target(e->v.Starred.value, del_targets);
+            }
```
```suggestion
            return _PyPegen_get_invalid_target(e->v.Starred.value, del_targets);
```

gvanrossum commented on Grammar/python.gram:
```diff
@@ -666,6 +676,25 @@ invalid_lambda_star_etc:
+
+# Comparison is used here, because the `a() in b` in `for a() in b: pass` is parsed
+# as a comparison. Thus, we only need to search the left side of the comparison
+# for invalid targets.
```
This seems dodgy. It inspired me to try this:
```
>>> for i < (): pass
Segmentation fault: 11
$ 
```

gvanrossum commented on Grammar/python.gram:
```diff
@@ -666,6 +676,25 @@ invalid_lambda_star_etc:
-    | a=star_expression &del_target_end {
-        RAISE_SYNTAX_ERROR_KNOWN_LOCATION(a, "cannot delete %s", _PyPegen_get_expr_name(a)) }
+invalid_with_target:
+    | ASYNC? 'with' expression 'as' a=expressions ':' {
```
What about multiple context managers? Compare these:
```
>>> with g() as 1: pass
  File "<stdin>", line 1
    with g() as 1: pass
                ^
SyntaxError: cannot assign to literal
>>> 
>>> with f() as a, g() as 1: pass
  File "<stdin>", line 1
    with f() as a, g() as 1: pass
                           ^
SyntaxError: invalid syntax
>>> 
```

gvanrossum commented on Parser/pegen/pegen.c:
```diff
@@ -2058,7 +2058,7 @@ _PyPegen_make_module(Parser *p, asdl_seq *a) {
 
 expr_ty
-_PyPegen_get_invalid_target(expr_ty e)
+_PyPegen_get_invalid_target(expr_ty e, int del_targets)
```
Can you please adopt a naming convention that makes it clear that this is effectively a bool? I was a bit baffled by this for a while. A venerable convention is to use names like `is_foo` or `has_bar`.

lysnikolaou replied:
Not sure, if `is_del` is okay, because the questions "What's del, the target, which target?" comes to mind. What do you think?

gvanrossum reviewed (commented):
Almost there!

lysnikolaou committed:
Update Grammar/python.gram

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

lysnikolaou committed:
Update Parser/pegen/pegen.c

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

lysnikolaou committed:
Address Guido's feedback

Correctly identify invalid targets, when there are multiple context
managers in a with statement, avoid SEGFAULT in invalid for targets
and improve paremeter naming in _PyPegen_get_invalid_target.

lysnikolaou committed:
Inline _PyPegen_get_invalid_for_target

pablogsal commented on Parser/pegen/pegen.h:
```diff
@@ -261,11 +261,23 @@ int _PyPegen_check_barry_as_flufl(Parser *);
 #define GET_INVALID_DEL_TARGET(e) _PyPegen_get_invalid_target(e, 1)
 
+Py_LOCAL_INLINE(expr_ty)
+_PyPegen_get_invalid_for_target(expr_ty e)
```
Given that this is mainly for error reporting (so no hot path), I am not sure if it makes sense to force the inlining, no?

lysnikolaou replied:
I liked the idea of having one main visitor, `_PyPegen_get_invalid_target`, which is a function and then build macros around it. This seemed too big for a macro and so I chose an inline function. Do we only use inline functions for speedup? I'm certainly no expert, so it's your call.

pablogsal replied:
> Do we only use inline functions for speedup?

Normally is done for performance or for symbol visibility (check the C99 inline visibility rules). I am happy with the function, I just surprised on why you added the `Py_LOCAL_INLINE`.

lysnikolaou replied:
Ok then! Reverting.

lysnikolaou committed:
Revert inlining

pablogsal committed:
Merge remote-tracking branch 'upstream/master' into targets-errors

gvanrossum reviewed (commented):
Hm... I poked another hole here. :-(
```
>>> for i, (f, g()) in 10: pass
Assertion failed: (e != NULL && e->kind == Compare_kind), function _PyPegen_get_invalid_for_target, file Parser/pegen/pegen.c, line 2110.
Abort trap: 6
```

lysnikolaou committed:
Fix searching for invalid for targets, by visiting Compare nodes in _PyPegen_get_invalid_target

lysnikolaou committed:
Call RAISE_SYNTAX_ERROR instead of returning NULL, when no invalid for target was found

pablogsal commented on Parser/pegen/pegen.c:
```diff
@@ -2058,7 +2058,7 @@ _PyPegen_make_module(Parser *p, asdl_seq *a) {
 
 expr_ty
-_PyPegen_get_invalid_target(expr_ty e, int is_del)
+_PyPegen_get_invalid_target(expr_ty e, int is_del, int is_for)
```
Given that `is_del` and `is_for` cannot be passed together I think this should be two functions or it should receive an enum or something like that.

lysnikolaou replied:
Done in https://github.com/python/cpython/pull/20106/commits/c1cf8e63562999fe2ad8fac6a327e04c49f17aef.

lysnikolaou committed:
Merge branch 'master' into targets-errors

lysnikolaou committed:
Replace is_del and is_for args with an enum

lysnikolaou committed:
Only return an invalid target, when the comparison is an In comparison

lysnikolaou committed:
Fix bug

lysnikolaou committed:
Merge branch 'master' into targets-errors

lysnikolaou committed:
Merge branch 'master' into targets-errors

lysnikolaou committed:
Minor refactoring

lysnikolaou committed:
Merge branch 'master' into targets-errors

lysnikolaou committed:
Merge branch 'master' into targets-errors

lysnikolaou committed:
Merge branch 'master' into targets-errors

pablogsal commented on Parser/pegen.h:
```diff
@@ -263,11 +263,21 @@ int _PyPegen_check_barry_as_flufl(Parser *);
+    FOR_TARGETS
+} TARGETS_TYPE;
+expr_ty _PyPegen_get_invalid_target(expr_ty e, TARGETS_TYPE targets_type);
+#define GET_INVALID_TARGET(e) _PyPegen_get_invalid_target(e, STAR_TARGETS)
```
Maybe we should surround `_PyPegen_get_invalid_target` in a `CHECK` to make sure this does not return NULL (and not crash in that case).

lysnikolaou committed:
Add CHECK* checks to GET_INVALID_*TARGET macros

pablogsal reviewed (approved):
After reviewing it again I have played with it a bit more (before the `CHECK` commit) and I was not able to break it in any way so I think we can land it now so more people interact with it and it this way we could detect any non-trivial problems if they exist.
